### DISCUSSION TITLE
Could this be used as an alternative to Typechain?

### DISCUSSION THREAD

**Discussion by johanneskares** [wagmi-dev/abitype#48] [General]
Great project! Thanks so much! Could this be used as an alternative to Typechain? (https://github.com/dethcrypto/TypeChain)

I love that I don't have to use a separate command line generator but can use the ABI directly to create types. Is this project being used to create an `ethers` contract interface?

**Comment by tmm** [comment]
Definitely! Here's an example of typing ethers.js `Contract` in `@wagmi/core` through [`getContract`](https://github.com/wagmi-dev/wagmi/blob/main/packages/core/src/actions/contracts/getContract.ts) and [TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgcwKYwMIQHYygQwGMYAaOVKQgJgAYBBAIQEk4BfOAMyghDgHIAAgHd8yEMAD0haKj4BYAFCLp2AM7wVeIvAC8KdFlwFiACgSK4cfABNrUVKtUAufjQAegMgI+JC1YBGwC4U1PTMPgqsAJTKOFrEAHQcAK7YxMA4qooSEpa5lgB6APyKQA) if you want to try it out.

**Reply by johanneskares** [reply]
oh wow! I was looking for exactly this! Thanks so much! Should be mentioned in the Readme somewhere! I didn't come across it in the documentation of either `wagmi` or `abitype`

**Reply by peetzweg** [reply]
Was looking for the same thing. Looks quite complex to use it with just ethers without `wagmi`. I like `wagmi`, but sometimes just want use plain `ethers`.

Looks like we need todo some of this magic with this [`Contract`](https://github.com/wagmi-dev/wagmi/blob/main/packages/core/src/actions/contracts/getContract.ts#L103) type, which is included in `wagmi`. Can we somehow make `abitype` usable with `ethers` without `wagmi`?

**Reply by tmm** [reply]
Would highly recommend using `getContract` from `@wagmi/core`.

[`getContract`](https://github.com/wagmi-dev/wagmi/blob/main/packages/core/src/actions/contracts/getContract.ts) is really just a wrapper around ethers `Contract` that only adds a few lines of additional runtime code. You can import `getContract` and use it without needing to set up a wagmi `Client`, etc. If you only use `getContract` and a modern tooling set up, it's likely that all other code from `@wagmi/core` will be eliminated from your final build.

**Reply by peetzweg** [reply]
`getContract` works great! I've adapted my library to generate outputs for ethers and abitype directly. It even splits the ABI in individual fragments to be imported individually to reduce bundle size.

https://github.com/peetzweg/abimate#usage-with-wagmi

I've added the script to generate these typescript files from hardhat/foundry outputs to the module as well now.

https://github.com/peetzweg/abimate#use-with-own-contracts-abimatecore

Could this maybe even be a great addition to the `wagmi` codebase?
